server/src/weaverbird/backends/utils/formula_builder.py

```diff
@@ -0,0 +1,566 @@
+                Out[2]: -1.7640186915887845
+        """
+        # those imports are local because they can be heavy and this function
+        # may not been called all the time !
+        import math
+        import operator
+
+        @lru_cache
+        def checkmath(x, *args):
+            if x not in [x for x in dir(math) if "__" not in x]:
+                raise InvalidFormula(f"Unknown mathematical func {x}()")
+            fun = getattr(math, x)
+            return fun(*args)
+
+        binOps = {
+            ast.Add: operator.add,
+            ast.Sub: operator.sub,
+            ast.Mult: operator.mul,
+            ast.Div: operator.truediv,
+            ast.Mod: operator.mod,
+            ast.Pow: operator.pow,
+            ast.Call: checkmath,
+            ast.BinOp: ast.BinOp,
+        }
+
+        unOps = {
+            ast.USub: operator.neg,
+            ast.UAdd: operator.pos,
+            ast.UnaryOp: ast.UnaryOp,
+        }
+
+        ops = tuple(binOps) + tuple(unOps)
+
+        tree = ast.parse(formula, mode='eval')
+
+        def _eval(node):
+            if isinstance(node, ast.Expr):
+                return _eval(node.value)
+            elif isinstance(node, ast.Expression):
+                return _eval(node.body)
+            elif isinstance(node, ast.Str):
+                return node.s
+            elif isinstance(node, ast.Num):
+                return node.value
+            elif isinstance(node, ast.Constant):
+                return node.value
+            elif isinstance(node, ast.BinOp):
+                if isinstance(node.left, ops):
+                    left = _eval(node.left)
+                else:
+                    left = getattr(node.left, 'value')
+                if isinstance(node.right, ops):
+                    right = _eval(node.right)
+                else:
+                    right = getattr(node.right, 'value')
+                return binOps[type(node.op)](left, right)  # type: ignore
+            elif isinstance(node, ast.UnaryOp):
+                if isinstance(node.operand, ops):
+                    operand = _eval(node.operand)
+                else:
+                    operand = getattr(node.operand, 'value')
+                return unOps[type(node.op)](operand)  # type: ignore
+            elif isinstance(node, ast.Call):
+                args = [_eval(x) for x in node.args]
+                r = checkmath(getattr(node.func, 'id'), *args)
+                return r
+            else:
+                raise InvalidFormula(f"Bad formula syntax, {type(node)}")
+
+        return _eval(tree)
+
+    @classmethod
+    def translate_formula(cls, step: FormulaStep) -> Any:
+        return formula
+
+    @classmethod
+    def sanitize_formula(cls, formula: str) -> Any:
+        return {}, formula
+
+    @classmethod
+    def build_formula_tree(
+        cls, formula: str | int | float | bool, quote_char: str | None = None
+    ) -> Any:
```

Comment:
Would that be clearer if that would return tree (i.e. the root the  (`Node`) and not `Any`?